pkg/build/registry/buildlog/rest.go

```diff
@@ -43,62 +48,97 @@ func (r RealPodControl) getPod(namespace, name string) (*kapi.Pod, error) {
-	// TODO: choose a more appropriate "try again later" status code, like 202
+	switch build.Status {
+	// Build has not launched, wait til it runs
+	case api.BuildStatusNew, api.BuildStatusPending:
+		if !buildLogOpts.Wait {
+			return nil, nil
+		}
+		err := r.waitForBuild(ctx, build)
+		if err != nil {
+			return nil, err
+		}
+	// An error occurred launching the build, return an error
+	case api.BuildStatusError:
+		return nil, fmt.Errorf("build %s is in an error state", build.Name)
+	}
+	// The container should be the default build container, so setting it to blank
 	buildPodName := buildutil.GetBuildPodName(build)
-	pod, err := r.PodControl.getPod(build.Namespace, buildPodName)
-	if err != nil {
-		return nil, nil, fielderrors.NewFieldNotFound("Pod.Name", buildPodName)
+	logOpts := &kapi.PodLogOptions{
+		Follow: buildLogOpts.Follow,
 	}
-
-	buildPodHost := pod.Spec.Host
-	buildPodNamespace := pod.Namespace
-	// Build will take place only in one container
-	buildContainerName := pod.Spec.Containers[0].Name
-
-	scheme, port, transport, err := r.ConnectionInfo.GetConnectionInfo(buildPodHost)
+	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, buildPodName, logOpts)
 	if err != nil {
-		return nil, nil, err
-	}
-
-	location := &url.URL{
-		Scheme: scheme,
-		Host:   net.JoinHostPort(buildPodHost, strconv.FormatUint(uint64(port), 10)),
-		Path:   fmt.Sprintf("/containerLogs/%s/%s/%s", buildPodNamespace, buildPodName, buildContainerName),
+		return nil, err
 	}
+	return &genericrest.LocationStreamer{
+		Location:    location,
+		Transport:   transport,
+		ContentType: "text/plain",
+		Flush:       buildLogOpts.Follow,
+	}, nil
+}
 
-	// Pod in which build take place can't be in the Pending or Unknown phase,
-	// cause no containers are present in the Pod in those phases.
-	if pod.Status.Phase == kapi.PodPending || pod.Status.Phase == kapi.PodUnknown {
-		return nil, nil, fielderrors.NewFieldInvalid("Pod.Status", pod.Status.Phase, "must be Running, Succeeded or Failed")
+func (r *REST) waitForBuild(ctx kapi.Context, build *api.Build) error {
+	fieldSelector := fields.Set{"name": build.Name}.AsSelector()
+	w, err := r.BuildRegistry.WatchBuilds(ctx, labels.Everything(), fieldSelector, build.ResourceVersion)
+	if err != nil {
+		return err
 	}
-
-	switch build.Status {
-	case api.BuildStatusRunning:
-		location.RawQuery = "follow=1"
-	case api.BuildStatusComplete, api.BuildStatusFailed:
-		// Do not follow the Complete and Failed logs as the streaming already finished.
-	default:
-		return nil, nil, fielderrors.NewFieldInvalid("build.Status", build.Status, "must be Running, Complete or Failed")
+	defer w.Stop()
+	done := make(chan struct{})
+	errchan := make(chan error)
+	go func(ch <-chan watch.Event) {
+		for event := range ch {
+			obj, ok := event.Object.(*api.Build)
+			if !ok {
+				errchan <- fmt.Errorf("event object is not a build: %#v", event.Object)
+				break
+			}
+			switch obj.Status {
```

Comment:
You're right. In both cases, I'd need to check for BuildStatusCancelled in addition to BuildStatusError. Thanks!